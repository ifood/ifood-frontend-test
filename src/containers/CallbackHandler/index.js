import React, { useEffect } from "react";
import queryString from "query-string";
import { useHistory } from "react-router-dom";
import { saveAuthItems } from "utils/auth";

export default function CallbackHandler() {
  const history = useHistory();
  useEffect(() => {
    const params = queryString.parse(window.location.search);
    if (params.error) {
      if (params.error === "access_denied") {
        return history.push("/intro?denied=true");
      }
      return history.push("/intro?error=true");
    }
    saveAuthItems();
    history.push("/");
  }, []);
  return null;
}
