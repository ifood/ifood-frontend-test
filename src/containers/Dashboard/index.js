import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

import { isAuthenticated } from "services/api";
import { UserActions } from "store/ducks/user";

import Loading from "components/core/Loading";
import Layout from "components/core/Layout";
import Auth from "./Auth";
import Playlists from "./Playlists";

const Dashboard = () => {
  const dispatch = useDispatch();

  const { isLoading, userName } = useSelector(({ user }) => user);
  const { isSuccess } = useSelector(({ auth }) => auth);

  useEffect(() => {
    isAuthenticated() && dispatch(UserActions.getUser());
  }, [dispatch, isSuccess]);

  return (
    <Loading isLoading={isLoading}>
      <Layout userName={userName}>
        {!isAuthenticated() ? <Auth /> : <Playlists />}
      </Layout>
    </Loading>
  );
};

export default Dashboard;
