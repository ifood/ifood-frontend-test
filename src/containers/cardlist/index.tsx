import React, { useContext } from 'react';
import { Card } from 'components/card';
import { PlayListContext } from 'store/play-list-store';

export const CardList: React.FC = () => {
  const playlistContext = useContext(PlayListContext);
  const playlist = playlistContext.state.playlist;

  return (
    <>
      {playlist.map((item: any) => (
        <Card>
          <span>{item.name}</span>
        </Card>
      ))}
    </>
  );
};
