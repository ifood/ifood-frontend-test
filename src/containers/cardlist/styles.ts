import styled from 'styled-components';
import colors from 'common/colors';
import breakpoints from 'common/breakpoints';

export const Wrapper = styled.div`
  background: ${colors.RED};
  padding: 140px 20px;
  margin-top: -120px;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
`;

export const Header = styled.header`
  height: 35px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-bottom: 1px solid #ff00004f;

  span {
    font-weight: 600;
  }
`;

export const Footer = styled.footer`
  height: 35px;
  display: flex;
  justify-content: center;
  align-items: center;
  padding-bottom: 3px;

  a {
    width: 100%;
    font-weight: 600;
    text-align: center;
    color: ${colors.RED};
    text-decoration: none;
    &:hover {
      color: ${colors.GRAY_LIGHT};
    }
  }

  @media (${breakpoints.LARGE}) {
    cursor: pointer;
  }
`;
