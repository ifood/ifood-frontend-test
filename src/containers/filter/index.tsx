import React, { useState } from 'react';

import { DateField } from 'components/date-field';
import { Select } from 'components/select';
import { Input } from 'components/input';
import { Button } from 'components/button';

//locale
//country
//timestamp
//limit
//offset

export const Filter: React.FC = () => {
  const [startDate, setStartDate] = useState(new Date());

  const handleDateChange = (date: Date) => {
    setStartDate(date);
  };

  return (
    <>
      <h1>Filter</h1>
      <DateField startDate={startDate} onChangeDate={(date) => handleDateChange(date)} />
      <Select />
      <Input />
      <Button />
    </>
  );
};
