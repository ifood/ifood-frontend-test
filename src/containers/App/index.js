import React from 'react'
import { Switch, Route } from 'react-router-dom'
import { ThemeProvider } from 'styled-components'

import HomePage from '../HomePage'
import NotFoundPage from '../NotFoundPage'
import PlaylistsPage from '../PlaylistsPage'

import Header from '../../components/Header'
import Footer from '../../components/Footer'

import { HOME_PATH } from './urls'
import './styles.css'
import { theme } from './theme'

function App() {
  return (
    <ThemeProvider theme={theme}>
      <div className="App">
        <Header />

        <div className="App-container">
          <Switch>
            <Route exact path="/" component={HomePage} />
            <Route path={HOME_PATH} component={PlaylistsPage} />
            <Route component={NotFoundPage} />
          </Switch>
        </div>

        <Footer />
      </div>
    </ThemeProvider>
  )
}

export default App
