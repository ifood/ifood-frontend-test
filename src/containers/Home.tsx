import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import { Dispatch } from 'redux';

import Filter from '../components/Filter';
import {
  operations as authOps,
  selectors as authSelectors,
} from '../ducks/auth';
import { operations as filterOps } from '../ducks/filter';
import { operations as playlistOps } from '../ducks/playlist';

interface IProps {
  token: string;
  getFilterConfig: () => Dispatch;
  listFeaturedPlaylists: (token: string) => Dispatch;
  signOut: () => Dispatch;
}

class Home extends PureComponent<IProps> {
  public componentDidMount() {
    this.props.getFilterConfig();
    this.props.listFeaturedPlaylists(this.props.token);
  }

  public render() {
    return (
      <div>
        <Filter onSignOut={this.handleSignOut} />
        <h1>Hi, I'm the list component</h1>
      </div>
    );
  }

  private handleSignOut = () => {
    this.props.signOut();
  };
}

const mapStateToProps = state => ({
  token: authSelectors.getToken(state),
});

const mapDispatchToProps = dispatch => ({
  getFilterConfig: () => dispatch(filterOps.getConfig()),
  listFeaturedPlaylists: (token: string) =>
    dispatch(playlistOps.listFeaturedPlaylists(token)),
  signOut: () => dispatch(authOps.signOut()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Home);
