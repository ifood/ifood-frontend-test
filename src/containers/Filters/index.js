import React, { useState, useEffect, useCallback, useMemo } from "react";
import { useSelector, useDispatch } from "react-redux";

import Loader from "components/Loader";
import useEffectUpdate from "hooks/useEffectUpdate";
import FiltersApiService from "services/filters";
import PlaylistsApiService from "services/playlists";
import { getFilterField } from "helpers/getFilterField";
import { filtersContainerData } from "constants/data/containers/Filters";

import {
  updatePlaylists,
  removePlaylists,
  removeFilteredPlaylists,
} from "store/modules/playlists/actions";

import * as S from "./styled";

const Filters = () => {
  const [fields, setFields] = useState([]);
  const [dataStart, setDataStart] = useState(new Date());
  const [params, setParams] = useState(false);
  const [limitSize, setLimitSize] = useState(false);
  const [offsetIsNegative, setOffsetIsNegative] = useState(false);
  const [generalError, setGeneralError] = useState(false);

  const dispatch = useDispatch();

  const token = useSelector((state) => state.authentication.token);

  useEffect(() => {
    const requestFilters = FiltersApiService();
    const { getFilters } = requestFilters;

    getFilters()
      .then((response) => {
        setFields(response);
      })
      .catch((error) => {
        return error;
      });
  }, []);

  const updateParams = useCallback((id, value) => {
    const filter = {
      [id]: value,
    };

    setParams((prevState) => {
      return { ...prevState, ...filter };
    });
  }, []);

  useEffectUpdate(() => {
    const requestPlaylists = PlaylistsApiService();
    const { getPlaylists } = requestPlaylists;

    getPlaylists(token, params)
      .then((data) => {
        dispatch(updatePlaylists(data));
        dispatch(removeFilteredPlaylists());
      })
      .catch(() => {
        setGeneralError(true);
        dispatch(removePlaylists());
      });
  }, [params]);

  const handleInputChange = useCallback(
    (e, id) => {
      const { value } = e.target;

      if (id === "offset" && Number(value) < 0) {
        return setOffsetIsNegative(true);
      }
      setOffsetIsNegative(false);

      if (id === "limit" && (Number(value) > 50 || Number(value) < 1)) {
        return setLimitSize(true);
      }
      setLimitSize(false);

      if (id === "country" && value === "en_US") {
        return updateParams(id, "US");
      }

      if (value !== "") {
        return updateParams(id, value);
      }
    },
    [updateParams]
  );

  const handleDateChange = useCallback(
    (date, id) => {
      const convertedDate = date.toISOString();

      updateParams(id, convertedDate);

      setDataStart(date);
    },
    [updateParams]
  );

  const fieldsSize = useMemo(() => {
    return fields.length === 0;
  }, [fields]);

  return (
    <S.Filters>
      <S.FiltersBox>
        {fieldsSize && <Loader />}

        {offsetIsNegative && (
          <S.FiltersValidation>
            {filtersContainerData.errors.offset}
          </S.FiltersValidation>
        )}

        {limitSize && (
          <S.FiltersValidation>
            {filtersContainerData.errors.limit}
          </S.FiltersValidation>
        )}

        {generalError && (
          <S.FiltersValidation>
            {filtersContainerData.errors.general}
          </S.FiltersValidation>
        )}

        {fields.map((item) => (
          <S.FiltersItem key={item.id} fieldType={item.id}>
            {getFilterField(
              item,
              (e) => handleInputChange(e, item.id),
              (date) => handleDateChange(date, item.id),
              dataStart
            )}
          </S.FiltersItem>
        ))}
      </S.FiltersBox>
    </S.Filters>
  );
};

export default Filters;
