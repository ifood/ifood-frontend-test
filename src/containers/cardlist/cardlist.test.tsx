import React from 'react';
import { render, screen } from '@testing-library/react';
import { CardList } from 'containers/cardlist';
import { axe, toHaveNoViolations } from 'jest-axe';

import { PlayListContext } from 'store/play-list-store';

expect.extend(toHaveNoViolations);

const list = [
  {
    external_urls: {
      spotify: 'https://open.spotify.com/playlist/37i9dQZF1DX5bjCEbRU4SJ'
    },
    id: '37i9dQZF1DX5bjCEbRU4SJ',
    images: [
      {
        height: null,
        url: 'https://i.scdn.co/image/ab67706f00000003278197087524cc094f86e82b',
        width: null
      }
    ],
    name: 'Calm Down'
  }
];

const wrapper = (playlist = [], filteredList = [], empty = false) =>
  render(
    <PlayListContext.Provider
      value={{
        state: {
          playlist: playlist,
          filteredList: filteredList,
          emptyFilterList: empty
        },
        dispatch: {
          setPlaylist: jest.fn(),
          setFilteredList: jest.fn(),
          setEmptyFilterList: jest.fn()
        }
      }}
    >
      <CardList />
    </PlayListContext.Provider>
  );

test('testing accessibility', async () => {
  const { container } = wrapper();
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});

test('renders component without errors', () => {
  //@ts-ignore
  const { container } = wrapper(list, list, false);

  expect(screen.getByText(/Calm Down/)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
