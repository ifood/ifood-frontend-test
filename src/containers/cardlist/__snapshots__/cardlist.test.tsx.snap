// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component without errors 1`] = `
<div>
  <div
    class="sc-dlnjPT ietYiT"
  >
    <div
      class="sc-bdnylx kIxwpW"
    >
      <header
        class="sc-hKFyIo excVIu"
      >
        <span>
          Calm Down
        </span>
      </header>
      <img
        alt="Imagem do album Calm Down"
        class="sc-gtssRu Svgfd"
        height="298px"
        src="https://i.scdn.co/image/ab67706f00000003278197087524cc094f86e82b"
      />
      <footer
        class="sc-eCApGN hiuAxn"
      >
        <a
          href="[object Object]"
        >
          Ou√ßa agora
        </a>
      </footer>
    </div>
  </div>
</div>
`;
